#878 - Add slaveId in MarathonTask and API to retrieve Mesos Master URL

janisz opened the pull request:
This PR prepare backend to expose sandbox URL in GUI https://github.com/mesosphere/marathon/issues/878. 

Sandbox URL is build from 3 pieces

1. Mesos master URL
2. SlaveID
3. TaskID

So I added SlaveID to MarathonTask, and added `/v2/mesos/` API to obtain current Mesos master.

kolloch commented on src/main/scala/mesosphere/marathon/api/v2/MesosResource.scala:
```diff
@@ -0,0 +1,24 @@
+    conf: MarathonConf) {
+  @GET
+  @Produces(Array(MediaType.APPLICATION_JSON))
+  def index(): Response = {
```
This is mostly already included in the `InfoResource`.

Just add the `"masterUrl"` field as `"mesos_master_url"`.

kolloch commented on src/main/scala/mesosphere/marathon/api/v2/json/Formats.scala:
```diff
@@ -71,7 +71,8 @@ trait Formats
       "stagedAt" -> (if (task.getStagedAt != 0) Timestamp(task.getStagedAt) else JsNull),
-      "version" -> task.getVersion
+      "version" -> task.getVersion,
+      "slaveId" -> task.getSlaveId.getValue
```
We cannot rely on the slaveId being present. Does this code correctly deal with that case?

kolloch commented on src/main/scala/mesosphere/marathon/api/v2/json/Formats.scala:
```diff
@@ -91,6 +92,11 @@ trait Formats
       withServicePorts
+
+    if (task.mesosTaskUrl.nonEmpty)
+      enrichedJson ++ Json.obj("mesosTaskUrl" -> task.mesosTaskUrl)
```
Json.obj should correctly deal with `None` values, so no need for an if-statement.

kolloch commented on src/main/scala/mesosphere/marathon/tasks/TaskUrlUtil.scala:
```diff
@@ -0,0 +1,20 @@
+import mesosphere.marathon.Protos.MarathonTask
+import mesosphere.util.state.{ FrameworkIdUtil, MesosMasterUtil }
+
+class TaskUrlUtil @Inject() (
```
I like that you split this out into a separate class!

Please add a short description.

kolloch commented on src/main/scala/mesosphere/marathon/tasks/TaskUrlUtil.scala:
```diff
@@ -0,0 +1,20 @@
+    for {
+      frameworkId <- frameworkIdUtil.fetch()
+    } yield s"$masterUrl/#/slaves/$slaveId/frameworks/${frameworkId.getValue}/executors/${task.getId}/"
+  }
```
Maybe add an example here as documentation?

kolloch commented on src/main/scala/mesosphere/marathon/tasks/TaskUrlUtil.scala:
```diff
@@ -0,0 +1,20 @@
+
+  def mesosTaskUrl(task: MarathonTask): Option[String] = {
+    val masterUrl = mesosMasterUtil.masterUrl
+    val slaveId = task.getSlaveId.getValue
```
treat this as an option. slaveId might not be set.

kolloch commented on src/test/scala/mesosphere/marathon/api/TaskKillerTest.scala:
```diff
@@ -50,8 +51,8 @@ class TaskKillerTest extends MarathonSpec
     val appId = PathId(List("my", "app"))
-    val task1 = MarathonTasks.makeTask("task-1", "host", ports = Nil, attributes = Nil, version = Timestamp.now())
-    val task2 = MarathonTasks.makeTask("task-2", "host", ports = Nil, attributes = Nil, version = Timestamp.now())
+    val task1 = MarathonTasks.makeTask("task-1", "host", ports = Nil, attributes = Nil, version = Timestamp.now(), slaveId = SlaveID.newBuilder.setValue("").build())
```
Please wrap these lines, they are getting really long.

kolloch commented:
BTW: This can actually also be used as a first step for https://github.com/mesosphere/marathon/issues/1745. Nice!

kolloch commented:
@janisz: My first impression is really good! Thanks for this PR!

I will continue my review but please also make sure that you have read this:

https://mesosphere.github.io/marathon/docs/contributing.html

It would be nice if you squashed your PR into one commit based on the current master.

kolloch commented:
Hi @janisz,

I finished my first review round. There is still one major weakness in your PR: With your code we hit Zookeeper really often for fetching the FrameworkId and the MasterInfo.

We get notified about FrameworkId/MasterInfo changes in `mesosphere.marathon.MarathonScheduler#registered` and `mesosphere.marathon.MarathonScheduler#reregistered`. We should save this info in main-memory and then use that for the MesosMasterUtil. For new code, we want to keep mutable shared state in actors.

We also need to be careful about relying on new fields like slaveId: We always have to deal gracefully with the case that the slaveId is not set (because the task was created before the upgrade).

We would really like this PR! So it would be great if you let us know if you are willing to continue working on this. If yes, we could also schedule a hangout when talking about a revision of this PR.

janisz commented:
@kolloch Thanks for review. I hope I'll fix all issues this week. 

In my first draft 87ccbd9b1dea20bd163c126a1ea598822547f163 I actually used `MarathonScheduler` to get notification about Mesos master changes but I was not sure if this was right approach.

kolloch commented:
Hi @janisz, I guess your first approach wasn't so bad, then.

You should make the field that you change `@volatile`, though, and make it an Option. Then make sure that you deal gracefully with the case that it is `None` (e.g. by simply not including the mesos task url in /v2/info).

I had a further discussion in my team right now: 

* We agreed to not export the `mesosTaskURL` for every task. Exporting the `mesos_master_url` in the info endpoint and the `slaveId` for every task should allow the Marathon UI to construct the URL itself.
* In addition, we would like a global configuration option to disable the automatic detection of the `mesos_master_url` and just set it to a constant configuration value.

The reason is that we are not sure how we expose information from Mesos in the Marathon UI in the future. Directly linking to the Mesos UI is often problematic because not all mesos master hosts might be reachable directly from the outside. Thus, we allow to overwrite this value to the publicly available mesos master URL (maybe load balanced). In the future, we might even proxy some things in Marathon to the Mesos API so that the browser never has to speak with the Mesos master directly.

Does that make sense?

kolloch commented:
@pierlo-upitup : Can you check if my above comment makes sense? Thanks.

janisz commented:
@kolloch So I should monitor MesosMaster with actor or with global variable?

kolloch commented:
@janisz Usually, we would use an actor for this but since it's only one variable, a @volatile Option[String] should be fine.

pierlo-upitup commented:
@kolloch 

Regarding: 
>  Exporting the mesos_master_url in the info endpoint and the slaveId for every task should allow the Marathon UI to construct the URL itself.

Sure, that's trivial.

janisz commented on src/main/scala/mesosphere/marathon/MarathonConf.scala:
```diff
@@ -15,6 +15,11 @@ trait MarathonConf extends ScallopConf with ZookeeperConf with IterativeOfferMat
+  lazy val mesosMasterUrl = opt[String]("mesos_master_url",
+    descr = "The host and port (e.g. \"http://mesos_host:5050\") of the Mesos master",
+    required = false,
+    noshort = true)
```
Should I validate URL?

kolloch replied:
I think that would be great but not a requirement for me.

Kornel commented on src/main/scala/mesosphere/marathon/MarathonModule.scala:
```diff
@@ -101,6 +101,17 @@ class MarathonModule(conf: MarathonConf, http: HttpConf, zk: ZooKeeperClient)
+  @Provides
+  @Singleton
+  def provideMesosLeaderInfo(): MesosLeaderInfo = {
+    if (conf.mesosMasterUrl.isDefined) {
```
getOrElse?

kolloch commented on src/main/scala/mesosphere/marathon/MarathonModule.scala:
```diff
@@ -101,6 +101,17 @@ class MarathonModule(conf: MarathonConf, http: HttpConf, zk: ZooKeeperClient)
+  @Singleton
+  def provideMesosLeaderInfo(): MesosLeaderInfo = {
+    if (conf.mesosMasterUrl.isDefined) {
+      ConstMesosLeaderInfo(Some(conf.mesosMasterUrl()))
```
I think `getOrElse` here would be great.

janisz replied:
I don't know how to use it here.

kolloch replied:
`getOrElse` doesn't fit, sorry. But match would:

```scala
    conf.mesosMasterUrl.get match {
      case someUrl @ Some(_) => ConstMesosLeaderInfo(someUrl)
      case None              => MutableMesosLeaderInfo
    }
```

kolloch commented on src/main/scala/mesosphere/util/state/MesosLeaderInfo.scala:
```diff
@@ -0,0 +1,30 @@
+  override def load(host: String, port: Int): Unit = {}
+}
+
+object MutableMesosLeaderInfo extends MesosLeaderInfo {
```
You asked me directly whether you use a global variable for this or not. For some reason, I misinterpreted the question. Otherwise I would have said: No, please don't! ;) Sorry about that.

So please make `MutableMesosLeaderInfo` a class instead of an object.

kolloch commented on src/main/scala/mesosphere/util/state/MesosLeaderInfo.scala:
```diff
@@ -0,0 +1,30 @@
+
+object MutableMesosLeaderInfo extends MesosLeaderInfo {
+
+  @volatile var leaderUrl: Option[String] = None
```
please make this `private[this]`

kolloch commented on src/main/scala/mesosphere/util/state/MesosLeaderInfo.scala:
```diff
@@ -0,0 +1,30 @@
+  */
+trait MesosLeaderInfo {
+
+  def currentLeaderUrl: Option[String]
```
Maybe document which URL it is exactly with an example. Does it include a trailing slash? etc.

kolloch commented on src/main/scala/mesosphere/util/state/MesosLeaderInfo.scala:
```diff
@@ -0,0 +1,30 @@
+
+  def currentLeaderUrl: Option[String]
+
+  def load(host: String, port: Int): Unit
```
Why not directly set the full URL here? I think that future Mesos versions will support HTTPS (maybe it is in 0.23 already?).

I also do not like the method name "load" so much. Maybe `onAutoDetectedMasterUrl`? That tries to express that the implementation can choose to use this information or not.

kolloch replied:
Or better: `onNewMasterInfo` and pass the `MasterInfo` as argument. The then the logic of constructing an URL from that is inside this object.

janisz replied:
`MasterInfo` has no information about protocol, that's why I decided to use host and port.

kolloch commented on docs/docs/rest-api.md:
```diff
@@ -1967,6 +1967,7 @@ Server: Jetty(8.y.z-SNAPSHOT)
         "local_port_max": 49151,
         "local_port_min": 32767,
         "master": "zk://localhost:2181/mesos",
+        "mesos_master_url": "http://mesos.vm:5050",
```
You should explain this better. Point out that it is automatically detected by default but you can override it with a certain config option.

kolloch replied:
Also add some documentation in command-line-flags.md.

kolloch commented:
Regarding your commit message:

* Your commit message should start with "Fixes #878 - "
* You should mention that you use this to auto detect the master url

kolloch commented on src/main/scala/mesosphere/marathon/MarathonScheduler.scala:
```diff
@@ -35,6 +35,7 @@ class MarathonScheduler @Inject() (
     taskTracker: TaskTracker,
     taskQueue: TaskQueue,
     frameworkIdUtil: FrameworkIdUtil,
+    mesosMasterUtil: MesosLeaderInfo,
```
call the variable `mesosLeaderInfo`

kolloch commented:
Hi @janisz, great patch! Thank you.

I made some comments, I hope that they make sense and are reasonably easy to fix.

I think there are still tests missing:

* That the DefaultTaskFactory copies the slaveId from the Offer into the MarathonTask
* For the slaveId in the Task JSON
* That MarathonScheduler.*registered callbacks call the appropriate method in `MesosLeaderInfo`
* For the InfoResource (admittedly there are none so far but we want to have tests for new code at least)

Not so important but easy to fix:

* If you initialize SlaveID in tests, I would prefer you to use `"slaveId"` or something like that instead of an empty string.

janisz commented:
@kolloch 
Thanks for review.

I hope I fixed most issues. I started with tests for `InfoResource` because I thought it will be easy, but I have no idea how properly mock configuration. With other tests I don't know if there are any good practices or convention. Tests in project look different (e.g. uses different assertions) so I'm not sure how to write them. 

Acording to
> That MarathonScheduler.*registered callbacks call the appropriate method in MesosLeaderInfo

I added check for it in `MarathonSchedulerTest` maybe it could be more explicit.

kolloch commented:
Hi @janisz, often, you do not need to mock config at all, you simply create it. See for example: https://github.com/mesosphere/marathon/blob/master/src/test/scala/mesosphere/marathon/core/launchqueue/impl/AppTaskLauncherActorTest.scala#L299-L300

If you need to supply special values for testing, you can do something similar to 
https://github.com/mesosphere/marathon/blob/master/src/test/scala/mesosphere/marathon/MarathonTestHelper.scala#L18-L44

(I would prefer you not to extend that but do that in your test suite)

kolloch commented:
This is a good patch. Thank you, @janisz .

You wrote above that you could make the tests more explicit. Do you still want to improve the testing there?

Otherwise this passes my bar for merging now :+1:

janisz commented:
@kolloch Thanks for hint, I've already found this helper but was wandering if there is any other way than creating full config to test `InfoResouce`.
I was thinking about extracting [this assertion] to separate test but it'll end up with same test twice (only different assertions). So I think there is nothing to improve. 

I see integration tests are failing. It's weird. Only change I made today is commit author because it's was wrong after rebase.

kolloch commented:
Hi @janisz, some of our integration tests are not reliable, unfortunately. I have restarted them.

janisz commented:
OK, build is green so I think it could be merge

kolloch commented:
Thank you @janisz !
